scripts/annotate_sam_with_bed_features.py

```diff
@@ -0,0 +1,328 @@
+        lowercase), and the tag is constructed accordingly
+        ("MIMAT0002849|6|4|11M3I11M|25|CTACAAAGGGAGGTAGCACTTTCTC").
+"""  # noqa: E501
+# pylint: enable=line-too-long
+
+import argparse
+from collections import defaultdict, namedtuple
+from pathlib import Path
+import sys
+from typing import Dict, Optional
+
+import pysam
+
+
+def parse_arguments():
+    """Command-line arguments parser."""
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument(
+        "-v",
+        "--version",
+        action="version",
+        version="%(prog)s 1.1.0",
+        help="Show program's version number and exit",
+    )
+    parser.add_argument(
+        "-b",
+        "--bed",
+        help=(
+            "Path to the BED file. This file must be the output of "
+            " a bedtools intersect call with `-a` as a GFF3 file and `-b as"
+            " a BAM file."
+        ),
+        type=Path,
+        required=True,
+    )
+    parser.add_argument(
+        "-s",
+        "--sam",
+        help="Path to the SAM input file.",
+        type=Path,
+        required=True,
+    )
+    parser.add_argument(
+        "-e",
+        "--extension",
+        help=(
+            "Number of nucleotides to adjust the feature coordinates: add to"
+            " the start and subtract from the end. Default: %(default)d."
+        ),
+        default=0,
+        type=int,
+    )
+    parser.add_argument(
+        "-S",
+        "--shift",
+        help=(
+            "Maximum allowed absoulte difference (in nucleotides) between the"
+            " alignment and feature coordinates at both ends. The tag is"
+            " added only if both 5' and 3' shifts are within this range"
+            " Default: %(default)d."
+        ),
+        default=0,
+        type=int,
+    )
+    parser.add_argument(
+        "--id",
+        help=(
+            "Feature attribute key to extract the indentifier (must be"
+            " lowercase). Default: %(default)s."
+        ),
+        default="name",
+        type=str,
+    )
+
+    return parser
+
+
+def attributes_dictionary(attr: str) -> Dict[str, str]:
+    """Convert an attribute string to a dictionary.
+
+    The attribute string is expected to be a semicolon-separated list of
+    key-value pairs. The function supports both '='-delimited and quoted
+    formats.
+
+    Args:
+        attr: The attribute string from a BED record.
+
+    Returns:
+        A dictionary with attribute keys in lowercase.
+    """
+    pairs = attr.split(";")
+
+    if len(pairs[0].split("=")) == 2:
+        attr_dict = {p.split("=")[0].lower(): p.split("=")[1] for p in pairs}
+    else:
+        attr_dict = {
+            p.split('"')[0].strip().lower(): p.split('"')[1] for p in pairs
+        }
+
+    return attr_dict
+
+
+def parse_intersect_output(
+    intersect_file: Path, ID: str = "name", extension: int = 0
```

Comment:
`ID` is a bad choice for a variable name, because:
- it violates the convention that variables are lower case and constants are upper case
- worse: it masks the built-in `id()` function